### ISSUE TITLE
chaoskube always kills the same pod

### ISSUE THREAD

**Issue by HaveFun83** [linki/chaoskube#197]
Hi 

currently upgrading from v0.15.1 to v0.19.0
```
NAME                              READY   STATUS    RESTARTS   AGE
chaoskube-demo-7f5ffd44db-djjcc   1/1     Running   0          10m
redis-demo-master-0               1/1     Running   0          8m29s
redis-demo-slave-0                1/1     Running   0          26s
redis-demo-slave-1                1/1     Running   0          7m39s
redis-demo-slave-2                1/1     Running   0          7m14s
```
and now always `redis-demo-slave-0` will be killed no random pod

```
time="2020-04-15T15:42:03Z" level=info msg="starting up" dryRun=false interval=2m0s version=v0.19.0
W0415 15:42:03.745799       6 client_config.go:543] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
time="2020-04-15T15:42:03Z" level=info msg="connected to cluster" master="https://10.96.0.1:443" serverVersion=v1.17.2
time="2020-04-15T15:42:03Z" level=info msg="setting pod filter" annotations= excludedPodNames="chaoskube-demo|redis-demo-master" includedPodNames="<nil>" labels= maxKill=1 minimumAge=0s namespaceLabels= namespaces=demo
time="2020-04-15T15:42:03Z" level=info msg="setting quiet times" daysOfYear="[]" timesOfDay="[16:00-07:00]" weekdays="[Friday Saturday Sunday]"
time="2020-04-15T15:42:03Z" level=info msg="setting timezone" location=UTC name=UTC offset=0
time="2020-04-15T15:46:04Z" level=info msg="terminating pod" name=redis-demo-slave-0 namespace=demo
time="2020-04-15T15:48:04Z" level=info msg="terminating pod" name=redis-demo-slave-0 namespace=demo
time="2020-04-15T15:50:04Z" level=info msg="terminating pod" name=redis-demo-slave-0 namespace=demo
time="2020-04-15T15:52:04Z" level=info msg="terminating pod" name=redis-demo-slave-0 namespace=demo
```
before upgrade chaoskube kills on of the `redis-demo-slave` pods randomly.

any ideas?

tia

**Comment by linki** [comment]
@HaveFun83 Interesting :thinking: 

Please run it with `--debug` which gives us some more information if it even has more candidates to pick from. It'll also show us the values of all flags.

**Comment by HaveFun83** [comment]
Log with from working v0.15.1
```
time="2020-04-17T10:54:07Z" level=debug msg="reading config" annotations= debug=true dryRun=false excludedDaysOfYear= excludedPodNames="chaoskube-demo|redis-demo-master" excludedTimesOfDay="16:00-07:00" excludedWeekdays="Sat,Sun" gracePeriod=-1s includedPodNames="<nil>" interval=2m0s kubeconfig= labels= logFormat=text master= metricsAddress=":8080" minimumAge=0s namespaceLabels= namespaces=demo timezone=UTC
time="2020-04-17T10:54:07Z" level=info msg="starting up" dryRun=false interval=2m0s version=v0.15.1
time="2020-04-17T10:54:07Z" level=debug msg="using cluster config" kubeconfig= master=
time="2020-04-17T10:54:07Z" level=info msg="connected to cluster" master="https://10.96.0.1:443" serverVersion=v1.17.2
time="2020-04-17T10:54:07Z" level=info msg="setting pod filter" annotations= excludedPodNames="chaoskube-demo|redis-demo-master" includedPodNames="<nil>" labels= minimumAge=0s namespaceLabels= namespaces=demo
time="2020-04-17T10:54:07Z" level=info msg="setting quiet times" daysOfYear="[]" timesOfDay="[16:00-07:00]" weekdays="[Saturday Sunday]"
time="2020-04-17T10:54:07Z" level=info msg="setting timezone" location=UTC name=UTC offset=0
time="2020-04-17T10:54:08Z" level=debug msg="found candidates" count=3
time="2020-04-17T10:54:08Z" level=info msg="terminating pod" name=redis-demo-slave-1 namespace=demo
time="2020-04-17T10:54:08Z" level=debug msg="calling deletePod endpoint" name=redis-demo-slave-1 namespace=demo terminator=DeletePod
time="2020-04-17T10:54:08Z" level=debug msg=sleeping...
time="2020-04-17T10:56:08Z" level=debug msg="found candidates" count=3
time="2020-04-17T10:56:08Z" level=info msg="terminating pod" name=redis-demo-slave-0 namespace=demo
time="2020-04-17T10:56:08Z" level=debug msg="calling deletePod endpoint" name=redis-demo-slave-0 namespace=demo terminator=DeletePod
time="2020-04-17T10:56:08Z" level=debug msg=sleeping...
time="2020-04-17T10:58:08Z" level=debug msg="found candidates" count=3
time="2020-04-17T10:58:08Z" level=info msg="terminating pod" name=redis-demo-slave-0 namespace=demo
time="2020-04-17T10:58:08Z" level=debug msg="calling deletePod endpoint" name=redis-demo-slave-0 namespace=demo terminator=DeletePod
time="2020-04-17T10:58:08Z" level=debug msg=sleeping...
time="2020-04-17T11:00:08Z" level=debug msg="found candidates" count=3
time="2020-04-17T11:00:08Z" level=info msg="terminating pod" name=redis-demo-slave-2 namespace=demo
time="2020-04-17T11:00:08Z" level=debug msg="calling deletePod endpoint" name=redis-demo-slave-2 namespace=demo terminator=DeletePod
time="2020-04-17T11:00:08Z" level=debug msg=sleeping...

```

Log from not working v0.19.0
```
time="2020-04-17T11:01:08Z" level=debug msg="reading config" annotations= debug=true dryRun=false excludedDaysOfYear= excludedPodNames="chaoskube-demo|redis-demo-master" excludedTimesOfDay="16:00-07:00" excludedWeekdays="Sat,Sun" gracePeriod=-1s includedPodNames="<nil>" interval=2m0s kubeconfig= labels= logFormat=text master= maxKill=1 metricsAddress=":8080" minimumAge=0s namespaceLabels= namespaces=demo slackWebhook= timezone=UTC
time="2020-04-17T11:01:08Z" level=info msg="starting up" dryRun=false interval=2m0s version=v0.19.0
time="2020-04-17T11:01:08Z" level=debug msg="using cluster config" kubeconfig= master=
W0417 11:01:08.166400       7 client_config.go:543] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
time="2020-04-17T11:01:08Z" level=info msg="connected to cluster" master="https://10.96.0.1:443" serverVersion=v1.17.2
time="2020-04-17T11:01:08Z" level=info msg="setting pod filter" annotations= excludedPodNames="chaoskube-demo|redis-demo-master" includedPodNames="<nil>" labels= maxKill=1 minimumAge=0s namespaceLabels= namespaces=demo
time="2020-04-17T11:01:08Z" level=info msg="setting quiet times" daysOfYear="[]" timesOfDay="[16:00-07:00]" weekdays="[Saturday Sunday]"
time="2020-04-17T11:01:08Z" level=info msg="setting timezone" location=UTC name=UTC offset=0
time="2020-04-17T11:01:09Z" level=debug msg="found victims" count=1
time="2020-04-17T11:01:09Z" level=info msg="terminating pod" name=redis-demo-slave-0 namespace=demo
time="2020-04-17T11:01:09Z" level=debug msg="calling deletePod endpoint" name=redis-demo-slave-0 namespace=demo terminator=DeletePod
time="2020-04-17T11:01:09Z" level=debug msg=sleeping...
time="2020-04-17T11:03:08Z" level=debug msg="found victims" count=1
time="2020-04-17T11:03:08Z" level=info msg="terminating pod" name=redis-demo-slave-0 namespace=demo
time="2020-04-17T11:03:08Z" level=debug msg="calling deletePod endpoint" name=redis-demo-slave-0 namespace=demo terminator=DeletePod
time="2020-04-17T11:03:08Z" level=debug msg=sleeping...
```

**Comment by linki** [comment]
I think the reason is [this](https://github.com/linki/chaoskube/blob/4fc706efe1a3dc80c89ca333840ac072e6177f8c/chaoskube/chaoskube.go#L444-L464).

This was done to prevent killing pods from the same replication group, such as Deployment, StatefulSet and so on. It's only needed when `--max-kill` is more than 1.

However, looking at the implementation it looks like it always picks the first pod from the group as the target.

**Comment by ravikumar2000** [comment]
Even chaoskube is restarting an not killing nay pods
Please find the logs
root@kmaster:~# helm install stable/chaoskube --generate-name
NAME: chaoskube-1587990114
LAST DEPLOYED: Mon Apr 27 17:51:58 2020
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
chaoskube is running and will kill arbitrary pods every 10m.

You can follow the logs to see what chaoskube does:

    POD=$(kubectl -n default get pods -l='app.kubernetes.io/instance=chaoskube-1587990114' --output=jsonpath='{.items[0].metadata.name}')
    kubectl -n default logs -f $POD

You are running in dry-run mode. No pod is actually terminated.
root@kmaster:~# helm install stable/chaoskube --generate-name --debug --set dryRun=false
install.go:159: [debug] Original chart version: ""
install.go:176: [debug] CHART PATH: /root/.cache/helm/repository/chaoskube-3.1.4.tgz

client.go:108: [debug] creating 1 resource(s)
NAME: chaoskube-1587990210
LAST DEPLOYED: Mon Apr 27 17:53:31 2020
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE: None
USER-SUPPLIED VALUES:
dryRun: false

COMPUTED VALUES:
affinity: {}
annotations: null
debug: false
dryRun: false
excludedDaysOfYear: null
excludedPodNames: null
excludedTimesOfDay: null
excludedWeekdays: null
gracePeriod: -1s
image: quay.io/linki/chaoskube
imageTag: v0.14.0
includedPodNames: null
interval: 10m
labels: null
logFormat: null
metrics:
  enabled: false
  port: 8080
  service:
    port: 8080
    type: ClusterIP
  serviceMonitor:
    additionalLabels: {}
    enabled: false
minimumAge: 0s
name: chaoskube
namespaces: null
nodeSelector: {}
podAnnotations: {}
podLabels: {}
priorityClassName: ""
rbac:
  create: false
  serviceAccountName: default
replicas: 1
resources: {}
timezone: UTC
tolerations: []

HOOKS:
MANIFEST:
---
# Source: chaoskube/templates/deployment.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: chaoskube-1587990210
  labels:
    app.kubernetes.io/name: chaoskube
    app.kubernetes.io/managed-by: "Helm"
    app.kubernetes.io/instance: "chaoskube-1587990210"
    helm.sh/chart: chaoskube-3.1.4
spec:
  replicas: 1
  selector:
    matchLabels:
      app.kubernetes.io/name: chaoskube
      app.kubernetes.io/instance: chaoskube-1587990210
  template:
    metadata:
      labels:
        app.kubernetes.io/name: chaoskube
        app.kubernetes.io/managed-by: "Helm"
        app.kubernetes.io/instance: "chaoskube-1587990210"
        helm.sh/chart: chaoskube-3.1.4
    spec:
      containers:
        - name: chaoskube
          image: quay.io/linki/chaoskube:v0.14.0
          args:
            - --interval=10m
            - --labels=
            - --annotations=
            - --namespaces=
            - --no-dry-run
            - --excluded-weekdays=
            - --excluded-times-of-day=
            - --excluded-days-of-year=
            - --timezone=UTC
            - --minimum-age=0s
            - --grace-period=-1s
            - --metrics-address=
          resources:
            {}
          securityContext:
            runAsNonRoot: true
            runAsUser: 65534
            readOnlyRootFilesystem: true
            capabilities:
              drop: ["ALL"]
      serviceAccountName: "default"

NOTES:
chaoskube is running and will kill arbitrary pods every 10m.

You can follow the logs to see what chaoskube does:

    POD=$(kubectl -n default get pods -l='app.kubernetes.io/instance=chaoskube-1587990210' --output=jsonpath='{.items[0].metadata.name}')
    kubectl -n default logs -f $POD
root@kmaster:~# POD=$(kubectl -n default get pods -l='app.kubernetes.io/instance=chaoskube-1587990210' --output=jsonpath='{.items[0].metadata.name}')
root@kmaster:~#
root@kmaster:~# kubectl -n default logs -f $POD
Error from server (NotFound): the server could not find the requested resource ( pods/log chaoskube-1587990210-795f6d7848-bs8dg)
root@kmaster:~# kubectl describe pods chaoskube-1587990210-795f6d7848-bs8dg -n default
Name:         chaoskube-1587990210-795f6d7848-bs8dg
Namespace:    default
Priority:     0
Node:         knode/10.0.3.15
Start Time:   Mon, 27 Apr 2020 17:53:31 +0530
Labels:       app.kubernetes.io/instance=chaoskube-1587990210
              app.kubernetes.io/managed-by=Helm
              app.kubernetes.io/name=chaoskube
              helm.sh/chart=chaoskube-3.1.4
              pod-template-hash=795f6d7848
Annotations:  <none>
Status:       Running
IP:           192.168.1.16
IPs:
  IP:           192.168.1.16
Controlled By:  ReplicaSet/chaoskube-1587990210-795f6d7848
Containers:
  chaoskube:
    Container ID:  docker://189430f9e1c73b8d8f91fa83202c39a5c6b090cfd5847684b06cc1bad1a9fc8c
    Image:         quay.io/linki/chaoskube:v0.14.0
    Image ID:      docker-pullable://quay.io/linki/chaoskube@sha256:74a8314513d94da26d407f29d2dba621ec9e607f5b2abbe07c3f6a521e00c7a4
    Port:          <none>
    Host Port:     <none>
    Args:
      --interval=10m
      --labels=
      --annotations=
      --namespaces=
      --no-dry-run
      --excluded-weekdays=
      --excluded-times-of-day=
      --excluded-days-of-year=
      --timezone=UTC
      --minimum-age=0s
      --grace-period=-1s
      --metrics-address=
    State:          Running
      Started:      Mon, 27 Apr 2020 17:54:46 +0530
    Last State:     Terminated
      Reason:       Error
      Exit Code:    1
      Started:      Mon, 27 Apr 2020 17:54:04 +0530
      Finished:     Mon, 27 Apr 2020 17:54:34 +0530
    Ready:          True
    Restart Count:  2
    Environment:    <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-gdsnt (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             True
  ContainersReady   True
  PodScheduled      True
Volumes:
  default-token-gdsnt:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-gdsnt
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason     Age               From               Message
  ----     ------     ----              ----               -------
  Normal   Scheduled  80s               default-scheduler  Successfully assigned default/chaoskube-1587990210-795f6d7848-bs8dg to knode
  Warning  BackOff    16s               kubelet, knode     Back-off restarting failed container
  Normal   Pulled     5s (x3 over 79s)  kubelet, knode     Container image "quay.io/linki/chaoskube:v0.14.0" already present on machine
  Normal   Created    5s (x3 over 78s)  kubelet, knode     Created container chaoskube
  Normal   Started    5s (x3 over 78s)  kubelet, knode     Started container chaoskube
root@kmaster:~# kubectl describe pods chaoskube-1587990210-795f6d7848-bs8dg -n default
Name:         chaoskube-1587990210-795f6d7848-bs8dg
Namespace:    default
Priority:     0
Node:         knode/10.0.3.15
Start Time:   Mon, 27 Apr 2020 17:53:31 +0530
Labels:       app.kubernetes.io/instance=chaoskube-1587990210
              app.kubernetes.io/managed-by=Helm
              app.kubernetes.io/name=chaoskube
              helm.sh/chart=chaoskube-3.1.4
              pod-template-hash=795f6d7848
Annotations:  <none>
Status:       Running
IP:           192.168.1.16
IPs:
  IP:           192.168.1.16
Controlled By:  ReplicaSet/chaoskube-1587990210-795f6d7848
Containers:
  chaoskube:
    Container ID:  docker://189430f9e1c73b8d8f91fa83202c39a5c6b090cfd5847684b06cc1bad1a9fc8c
    Image:         quay.io/linki/chaoskube:v0.14.0
    Image ID:      docker-pullable://quay.io/linki/chaoskube@sha256:74a8314513d94da26d407f29d2dba621ec9e607f5b2abbe07c3f6a521e00c7a4
    Port:          <none>
    Host Port:     <none>
    Args:
      --interval=10m
      --labels=
      --annotations=
      --namespaces=
      --no-dry-run
      --excluded-weekdays=
      --excluded-times-of-day=
      --excluded-days-of-year=
      --timezone=UTC
      --minimum-age=0s
      --grace-period=-1s
      --metrics-address=
    State:          Running
      Started:      Mon, 27 Apr 2020 17:54:46 +0530
    Last State:     Terminated
      Reason:       Error
      Exit Code:    1
      Started:      Mon, 27 Apr 2020 17:54:04 +0530
      Finished:     Mon, 27 Apr 2020 17:54:34 +0530
    Ready:          True
    Restart Count:  2
    Environment:    <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-gdsnt (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             True
  ContainersReady   True
  PodScheduled      True
Volumes:
  default-token-gdsnt:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-gdsnt
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason     Age                From               Message
  ----     ------     ----               ----               -------
  Normal   Scheduled  85s                default-scheduler  Successfully assigned default/chaoskube-1587990210-795f6d7848-bs8dg to knode
  Warning  BackOff    21s                kubelet, knode     Back-off restarting failed container
  Normal   Pulled     10s (x3 over 84s)  kubelet, knode     Container image "quay.io/linki/chaoskube:v0.14.0" already present on machine
  Normal   Created    10s (x3 over 83s)  kubelet, knode     Created container chaoskube
  Normal   Started    10s (x3 over 83s)  kubelet, knode     Started container chaoskube

**Comment by ravikumar2000** [comment]
Every 2.0s: kubectl get deployments,pods --all-namespaces                                                               kmaster: Mon Apr 27 17:58:26 2020

NAMESPACE     NAME                                   READY   UP-TO-DATE   AVAILABLE   AGE
default       deployment.apps/apache                 0/1     1            0           2d7h
default       deployment.apps/chaoskube-1587990114   0/1     1            0           6m28s
default       deployment.apps/chaoskube-1587990210   0/1     1            0           4m55s
default       deployment.apps/nginx                  2/2     2            2           2d6h
kube-system   deployment.apps/coredns                2/2     2            2           2d7h
more-apps     deployment.apps/chaoskube              0/1     1            0           2d
more-apps     deployment.apps/ghost                  2/2     2            2           2d6h

NAMESPACE     NAME                                        READY   STATUS             RESTARTS   AGE
default       pod/apache-8454694d99-xgvrl                 0/1     ImagePullBackOff   0          2d7h
default       pod/chaoskube-1587990114-79c757f5cb-q8b8j   0/1     CrashLoopBackOff   5          6m28s
default       pod/chaoskube-1587990210-795f6d7848-bs8dg   0/1     CrashLoopBackOff   4          4m55s
default       pod/nginx-5ccf85b585-cv8m5                  1/1     Running            0          2d6h
default       pod/nginx-5ccf85b585-ws4xj                  1/1     Running            0          2d6h
kube-system   pod/coredns-66bff467f8-jnx5b                1/1     Running            0          2d7h
kube-system   pod/coredns-66bff467f8-s9qp5                1/1     Running            0          2d7h
kube-system   pod/etcd-kmaster                            1/1     Running            0          2d7h
kube-system   pod/kube-apiserver-kmaster                  1/1     Running            0          2d7h
kube-system   pod/kube-controller-manager-kmaster         1/1     Running            0          2d7h
kube-system   pod/kube-flannel-ds-amd64-6sd7v             1/1     Running            0          2d7h
kube-system   pod/kube-flannel-ds-amd64-9gg8w             1/1     Running            2          2d7h
kube-system   pod/kube-flannel-ds-amd64-bqvbg             1/1     Running            0          2d7h
kube-system   pod/kube-proxy-58zlr                        1/1     Running            0          2d7h
kube-system   pod/kube-proxy-dp6xb                        1/1     Running            0          2d7h
kube-system   pod/kube-proxy-vm75r                        1/1     Running            0          2d7h
kube-system   pod/kube-scheduler-kmaster                  1/1     Running            0          2d7h
more-apps     pod/chaoskube-56998c669c-kp5rx              0/1     CrashLoopBackOff   16         2d
more-apps     pod/ghost-588cb7bd9f-746bg                  1/1     Running            0          2d6h
more-apps     pod/ghost-588cb7bd9f-9qtm6                  1/1     Running            0          2d6h

**Comment by linki** [comment]
Fixed in https://github.com/linki/chaoskube/pull/203.

Thanks @HaveFun83 for reporting this.

**Comment by linki** [comment]
This is fixed in version [v0.20.0](https://github.com/linki/chaoskube/releases/tag/v0.20.0).
